lib/charms/sdcore_amf/v0/fiveg_n2.py

```diff
@@ -0,0 +1,326 @@
+
+    HOST = "amf"
+    PORT = 38412
+    IP_ADDRESS = "192.168.70.132"
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.n2_provider = N2Provides(self, "fiveg-n2")
+        self.framework.observe(
+            self.on.fiveg_n2_relation_joined, self._on_fiveg_n2_relation_joined
+        )
+
+    def _on_fiveg_n2_relation_joined(self, event: RelationJoinedEvent):
+        if self.unit.is_leader():
+            self.n2_provider.set_n2_information(
+                amf_ip_address=self.IP_ADDRESS,
+                amf_hostname=self.HOST,
+                amf_port=self.PORT,
+            )
+
+
+if __name__ == "__main__":
+    main(DummyFivegN2ProviderCharm)
+```
+
+"""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3bb439930da24fd09631af74f70ea394"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft push-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+from typing import Dict
+import logging
+from typing import Optional
+
+from interface_tester.schema_base import DataBagSchema  # type: ignore[import]
+from ops.charm import CharmBase, CharmEvents, RelationChangedEvent
+from pydantic import BaseModel, ValidationError
+from ops.model import Relation
+from ops.framework import EventBase, EventSource, Handle, Object
+from pydantic import BaseModel, Field, ValidationError, IPvAnyAddress
+
+logger = logging.getLogger(__name__)
+"""Schemas definition for the provider and requirer sides of the `fiveg_n2` interface.
+It exposes two interfaces.schema_base.DataBagSchema subclasses called:
+- ProviderSchema
+- RequirerSchema
+Examples:
+    ProviderSchema:
+        unit: <empty>
+        app: {
+            "amf_ip_address": "192.168.70.132"
+            "amf_hostname": "amf",
+            "amf_port": 38412
+        }
+    RequirerSchema:
+        unit: <empty>
+        app:  <empty>
+"""
+class ProviderAppData(BaseModel):
+    """Provider app data for fiveg_n2."""
+    amf_ip_address: IPvAnyAddress = Field(
+        description="IP Address to reach the AMF's N2 interface.",
+        examples=["192.168.70.132"]
+    )
+    amf_hostname: str = Field(
+        description="Hostname to reach the AMF's N2 interface.",
+        examples=["amf"]
+    )
+    amf_port: int = Field(
+        description="Port to reach the AMF's N2 interface.",
+        examples=[38412]
+    )
+
+class ProviderSchema(DataBagSchema):
+    """Provider schema for fiveg_n2."""
+    app: ProviderAppData
+
+def data_is_valid(data: dict) -> bool:
+    """Returns whether data is valid.
+
+    Args:
+        data (dict): Data to be validated.
+
+    Returns:
+        bool: True if data is valid, False otherwise.
+    """
+    try:
+        ProviderSchema(app=data)
+        return True
+    except ValidationError as e:
+        logger.error("Invalid data: %s", e)
+        return False
+
```

Comment:
I think it would be useful to add the lib path to tox paths. I see quite a few linting problems here.

Reply:
Thanks, added.